modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/reducer/PageStream.java

```diff
@@ -0,0 +1,277 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache.query.reducer;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.UUID;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.internal.processors.cache.query.GridCacheQueryAdapter;
+import org.apache.ignite.internal.util.typedef.internal.U;
+
+/**
+ * Iterator over pages stream. Pages are stored in a queue. After polling a queue try load a new page instead of it.
+ */
+class PageStream<R> {
+    /** Queue of data of results pages. */
+    protected final Queue<Collection<R>> queue = new LinkedList<>();
+
+    /** Iterator over current page. */
+    private Iterator<R> iter;
+
+    /**
+     * Dynamic collection of nodes that run this query. If a node finishes this query then remove it from this colleciton.
+     */
+    private final Collection<UUID> subgrid;
+
+    /**
+     * List of nodes that respons with cache query result pages. This collection should be cleaned before sending new
+     * cache query request.
+     */
+    private final Collection<UUID> rcvd = new HashSet<>();
```

Comment:
This class has a child NodeStream, which is per-node stream.
What semantic 'subgrid', 'queue' and 'rcvd' will have in a child class?

Reply:
The semantic is the same. We still have a `subgrid` even it's a single node, and it also has to track `rcvd` to avoid duplicate messages. I think we should reuse this code instead of implementing separate logic for single node stream, that actually will be pretty the same.

Reply:
why do you decide to split reducer's logic between actual reducer class and this one?

Reply:
Why don't use NodeStream only and merge results at higher level?
Maybe NodeStreams can even share same queue for the results.